cmd/memtierd/main.go

```diff
@@ -0,0 +1,397 @@
+	for scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Fields(line)
+		if fields[0] == "cgroup2" {
+			cgroupMountPoint = fields[1]
+			break
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Fatalf("failed to read /proc/mounts: %v", err)
+	}
+	findCmd := exec.Command("find", cgroupMountPoint, "-type", "d", "-wholename", fullPath)
+
+	fullCgroupPath, err := findCmd.Output()
+	if err != nil {
+		log.Fatalf("failed to run find command: %v", err)
+	}
+
+	log.Printf("Cgroup path: %s", string(fullCgroupPath))
+
+	return fullCgroupPath
+}
+
+func editMemtierdConfig(fullCgroupPath []byte, podName string, containerName string, template string) (string, string, string) {
+	templatePath := fmt.Sprintf("/templates/%s", template)
+	yamlFile, err := ioutil.ReadFile(templatePath)
+	if err != nil {
+		log.Fatalf("Error reading YAML file: %v\n", err)
+	}
+
+	// Create pod directory if it doesn't exist
+	podDirectory := fmt.Sprintf("/tmp/memtierd/%s", podName)
+	if err := os.MkdirAll("/host"+podDirectory, 0755); err != nil {
+		log.Fatalf("Error creating directory: %v", err)
+	}
+
+	outputFilePath := fmt.Sprintf("%s/memtierd.%s.output", podDirectory, containerName)
+
+	var memtierdConfig MemtierdConfig
+	err = yaml.Unmarshal(yamlFile, &memtierdConfig)
+	if err != nil {
+		log.Fatalf("Error unmarshaling YAML: %v\n", err)
+	}
+
+	fullCgroupPathString := string(fullCgroupPath)
+
+	// Edit the Policy and Routine configs
+	policyConfigFieldString := string(memtierdConfig.Policy.Config)
+	policyConfigFieldString = strings.Replace(policyConfigFieldString, "$CGROUP2_ABS_PATH", fullCgroupPathString, 1)
+
+	// Loop through the routines
+	for i := 0; i < len(memtierdConfig.Routines); i++ {
+		routineConfigFieldString := string(memtierdConfig.Routines[i].Config)
+		routineConfigFieldString = strings.Replace(routineConfigFieldString, "$MEMTIERD_SWAP_STATS_PATH", outputFilePath, 1)
+		memtierdConfig.Routines[i].Config = routineConfigFieldString
+	}
+
+	memtierdConfig.Policy.Config = policyConfigFieldString
+
+	out, err := yaml.Marshal(&memtierdConfig)
+	if err != nil {
+		log.Fatalf("Error marshaling YAML: %v\n", err)
+	}
+
+	configFilePath := fmt.Sprintf("/host"+podDirectory+"/%s.yaml", containerName)
+	err = ioutil.WriteFile(configFilePath, out, 0644)
+	if err != nil {
+		log.Fatalf("Error writing YAML file: %v\n", err)
+	}
+	log.Infof("YAML file successfully modified.")
+
+	return podDirectory, outputFilePath, configFilePath
+}
+
+func startMemtierd(podName string, containerName string, podDirectory string, outputFilePath string, configFilePath string) {
+	log.Infof("Starting Memtierd")
+
+	outputFile, err := os.OpenFile("/host"+outputFilePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		fmt.Printf("Failed to open output file: %v\n", err)
+	}
+
+	// Create the command and write its output to the output file
+	cmd := exec.Command("memtierd", "-c", "", "-config", configFilePath)
+	cmd.Stdout = outputFile
+	cmd.Stderr = outputFile
```

Comment:
Maybe it's okay to start like this. But in the long run wouldn't it be better to refactor memtierd so that it offers an easily consumable interface for importing and running it as part of other binaries (typically on a dedicated goroutine) ?

We did something similar a while ago just to eliminate most of the duplicated boilerplate code in our various plugins which set up resource management. I think it was commit b62f866e04f0d7ef81e8f2ad9485f6ce740a83cb.